🔴 🔴 🔴  Added `segmentation maps` support for DPT image processor

simonreise opened the pull request:
# Added `segmentation maps` support for DPT image processor

Most of image processors for vision models that support semantic segmentation task accept `images` and `segmentation_maps` as inputs, but for some reason DPT image processor does not process segmentation maps, only images. This PR can make code that one uses for training or evaluation of semantic segmentation models more reusable, as now DPT image processor can process segmentation maps as most of other image processors do.

I also added `do_reduce_labels` arg because other image processors that support segmentation masks use it.

I added two new tests: one that tests segmentation_masks support and one that tests if do_reduce_labels work as expected.

Most of the code is adapted from BEIT image processor.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@amyeroberts, @qubvel

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

simonreise committed:
Added `segmentation_maps` support for DPT image processor

simonreise committed:
Added `segmentation_maps` support for DPT image processor

simonreise committed:
Added tests for dpt image processor

simonreise committed:
Added tests for dpt image processor

LysandreJik commented:
cc @molbap as well in case bandwidth permits

molbap commented on tests/models/dpt/test_image_processing_dpt.py:
```diff
@@ -138,3 +168,124 @@ def test_keep_aspect_ratio(self):
+
+    def test_call_segmentation_maps(self):
+        # Initialize image_processing
+        image_processing = self.image_processing_class(**self.image_processor_dict)
```
nit, `image_processor` would be better

simonreise replied:
Renamed `image_processing` to `image_processor`. Should I also rename it in the other tests?

molbap commented on src/transformers/models/dpt/image_processing_dpt.py:
```diff
@@ -426,6 +458,51 @@ def preprocess(
+            # Add channel dimension if missing - needed for certain transformations
+            if segmentation_maps[0].ndim == 2:
+                added_channel_dim = True
+                segmentation_maps = [segmentation_map[None, ...] for segmentation_map in segmentation_maps]
+                input_data_format = ChannelDimension.FIRST
+            else:
+                added_channel_dim = False
+                if input_data_format is None:
+                    input_data_format = infer_channel_dimension_format(segmentation_maps[0], num_channels=1)
+
+            if do_reduce_labels:
+                segmentation_maps = [self.reduce_label(segmentation_map) for segmentation_map in segmentation_maps]
+
+            if do_resize:
+                segmentation_maps = [
+                    self.resize(
+                        image=segmentation_map,
+                        size=size,
+                        resample=resample,
+                        keep_aspect_ratio=keep_aspect_ratio,
+                        ensure_multiple_of=ensure_multiple_of,
+                        input_data_format=input_data_format,
+                    )
+                    for segmentation_map in segmentation_maps
+                ]
+
+            if do_pad:
+                segmentation_maps = [
+                    self.pad_image(
+                        image=segmentation_map, size_divisor=size_divisor, input_data_format=input_data_format
+                    )
+                    for segmentation_map in segmentation_maps
+                ]
+
+            # Remove extra channel dimension if added for processing
+            if added_channel_dim:
+                segmentation_maps = [segmentation_map.squeeze(0) for segmentation_map in segmentation_maps]
+            segmentation_maps = [segmentation_map.astype(np.int64) for segmentation_map in segmentation_maps]
+
+            data["labels"] = segmentation_maps
```
Perfect - if there isn't any difference with Beit, can this be wrapped in a `_preprocess_segmentation_map()` method in a loop, that can be flagged as `# Copied from ...` the beit image processor?

simonreise replied:
Wrapped segmentation map preprocessing code to `_preprocess_segmentation_map()`, and also moved image preprocessing to separate `_preprocess_image()` function and general preprocessing functionality to `_preprocess()` function.

molbap reviewed (commented):
LGTM - just a small refactor of the method to be more aligned with existing models!

simonreise committed:
Moved preprocessing into separate functions

simonreise committed:
Moved preprocessing into separate functions

simonreise commented:
Could you please re-review the pull request? In the last commit I made all the changes you asked for: wrapped segmentation map preprocessing code to separate functions, added comments and renamed a variable in tests. Do I need to make any other changes to the code?

molbap commented:
hey @simonreise , will review in a moment, we were all at a team gathering last week hence the inactivity. On my radar!

ArthurZucker commented:
@molbap you are fobidden to work for this week 🤣 go and rest, @yonigozlan will have a look! 🤗

qubvel commented:
friendly ping @yonigozlan

yonigozlan commented on src/transformers/models/dpt/image_processing_dpt.py:
```diff
@@ -275,10 +282,162 @@ def _get_pad(size, size_divisor):
 
         return pad(image, ((pad_size_left, pad_size_right), (pad_size_top, pad_size_bottom)), data_format=data_format)
 
+    def reduce_label(self, label: ImageInput) -> np.ndarray:
+        label = to_numpy_array(label)
+        # Avoid using underflow conversion
+        label[label == 0] = 255
+        label = label - 1
+        label[label == 254] = 255
+        return label
```
This seems to be fully copied from beit image processor, you should add a `# Copied from` statement above if that's the case :)

simonreise replied:
Done

yonigozlan commented on src/transformers/models/dpt/image_processing_dpt.py:
```diff
@@ -275,10 +282,162 @@ def _get_pad(size, size_divisor):
+        segmentation_map = segmentation_map.astype(np.int64)
+        return segmentation_map
+
+    def __call__(self, images, segmentation_maps=None, **kwargs):
+        # Overrides the `__call__` method of the `Preprocessor` class such that the images and segmentation maps can both
+        # be passed in as positional arguments.
+        return super().__call__(images, segmentation_maps=segmentation_maps, **kwargs)
```
Same here for adding a `# Copied from`, and same for all the other methods copied from beit as well.

yonigozlan commented on src/transformers/models/dpt/image_processing_dpt.py:
```diff
@@ -275,10 +282,162 @@ def _get_pad(size, size_divisor):
     def preprocess(
         self,
         images: ImageInput,
+        segmentation_maps: Optional[ImageInput] = None,
```
This is a bit tricky as it could be a breaking change, if some users use `do_resize` etc. as args and not kwargs. However this would not be good practice, and I don't see any way of adding `segmentation_maps` processing without breaking BC. I'll let a core maintainer give the green light on this or not.

yonigozlan commented on tests/models/dpt/test_image_processing_dpt.py:
```diff
@@ -138,3 +168,124 @@ def test_keep_aspect_ratio(self):
         pixel_values = image_processor(image, return_tensors="pt").pixel_values
 
         self.assertEqual(list(pixel_values.shape), [1, 3, 512, 672])
+
+    def test_call_segmentation_maps(self):
```
Same as before about the # Copied from, if this and the following tests are indeed fully copied from beit

yonigozlan reviewed (commented):
Thanks for working on this!
Looks good to me, and something that would be useful to have.
Just left some comments about adding `# Copied from` statements where needed, and about breaking backward compatibility (mainly intended for a core maintainer)

simonreise committed:
Added # Copied from statements

simonreise committed:
Added # Copied from statements

yonigozlan commented:
Thanks, but the `# Copied from` statement must be placed above the function definition. You can refer to other parts of the library to see how it's done.
This is not just for information purposes; it enables the `make fix-copies` CLI command to propagate any modifications in the original function to its copied versions.

After making the required changes, you can ensure everything is in order by running the `make fixup` command.

simonreise committed:
Fixed # Copied from statements

simonreise committed:
Fixed # Copied from statements

yonigozlan commented:
Thanks for iterating! you just have to rebase on main and check that the tests are still passing, then LGTM!

simonreise committed:
Merge branch 'segmentation-maps-for-dpt-image-processor' of https://github.com/simonreise/transformers into segmentation-maps-for-dpt-image-processor

simonreise committed:
Merge branch 'main' into segmentation-maps-for-dpt-image-processor

simonreise committed:
Merge branch 'main' into segmentation-maps-for-dpt-image-processor

yonigozlan commented:
Everything looks good to me, pinging @ArthurZucker for a final review.
There's a potential breaking change @ArthurZucker, see [this comment ](https://github.com/huggingface/transformers/pull/34345/files#r1889587930)

simonreise committed:
Merge branch 'main' into segmentation-maps-for-dpt-image-processor

simonreise commented:
There was a merge conflict that appeared after #35439 was merged into main. So I also changed the order of `do_rescale` and `is_scaled_image` in the code

ArthurZucker reviewed (approved):
Don't mind breaking this, but we need to add a 🔴 on the PR name to sort it out when releasing! 🤗
